2217 only view submission needed

elipe17 opened the pull request:
## Summary of Changes
- Made `fileUploadSections` redux state to be accessed across different components
- Updated setStt action to get the number of sections the stt needs to submit
- Added tests to cover a valid and invalid stt
- Updated backend Stt view to be a viewset to allow custom actions
Pull request closes #2217 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in as ofa system admin
2. Change your stt on the datafiles tab and verify you see three or four sections to submit (compare Alabama, California, and any tribe)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] SSP/TANF radio button continues to drive updates to Reports section below search button
- [x] Users will not longer see all Sections 1-4 but instead headers based on backend models
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if **necessary**

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Updated STTApiView to STTApiViewSet

elipe17 committed:
- overriding retrieve method to key off of stt name instead of pk

elipe17 committed:
- MVP of limiting submission boxes based on what an stt should be submitting

elipe17 committed:
- Fix logic to use correct section index.
- need to sort uploads by section index

elipe17 committed:
- Updated map logic to correctly order the sections

elipe17 committed:
- linting

elipe17 committed:
- update doc string

elipe17 committed:
- remove object destructuring
- add missing state

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2217-only-view-submission-needed

elipe17 committed:
- update ternary
- Add undefined checks

elipe17 committed:
- add missing expected values

elipe17 committed:
- add default to missing state in tests

elipe17 committed:
- fix reducer tests

elipe17 committed:
- Updated submission history tests with correct state

elipe17 committed:
- Fix upload report tests

elipe17 committed:
- Correct logic if reponse is null

elipe17 committed:
- Updated action to use correct return type when api request fails
- Updated tests

elipe17 committed:
- remove unnecessary branch

elipe17 committed:
- Update tests to reverse based on viewset

elipe17 committed:
- add test for file extenstion reducer

elipe17 committed:
- linting

elipe17 committed:
- Added error handling for empty stt

elipe17 committed:
- add test for empty stt

elipe17 committed:
- Update logic to aviod extra conditional

elipe17 commented:
- Update api endpoint to only return a boolean since we know all STTs will submit 3 or 4 sections. Just need to check which one.
- Test updating `pk` in view to `name` for readability

elipe17 committed:
- Moved section logic mostly to the backend
- Updated the frontend to leverage backend work

elipe17 committed:
- Updated logic on components

elipe17 committed:
- add empty stt check

jtimpe commented on tdrs-frontend/src/actions/reports.js:
```diff
@@ -269,8 +270,20 @@ export const SET_SELECTED_YEAR = 'SET_SELECTED_YEAR'
+export const setStt = (stt) => async (dispatch) => {
+  var response = undefined
+  if (stt !== '') {
+    const URL = `${process.env.REACT_APP_BACKEND_URL}/stts/${stt}/num_sections`
```
i still think there's an opportunity to eliminate this api call somehow. `SET_SELECTED_STT` isn't dispatched until the "Search" button is pressed on the datafile page - so not only does it have to reach out for the submitted datafiles/submission history, a second api call needs to be made just for the number of sections

`SET_STTS` is dispatched whenever the frontend retrieves the stt list
![image](https://github.com/user-attachments/assets/5a50247f-82c7-4992-bd18-0d27e8cb04f8)

if each of the stts in the list could have `num_sections`, i think we could avoid this extra api request

jtimpe replied:
i think that could be as simple as modifying the `STTSerializer` on the backend to have a new generated field

elipe17 replied:
I see what you're saying, I think that is a better idea! I will make that change!

elipe17 committed:
- Updated stt serializer to include the number of sections an stt submits
- Reverted code to hit `num_sections` endpoint
- Reverted tests
- Updated components to use new field in stt state

elipe17 committed:
- remove async
- correct error

jtimpe reviewed (approved):
Good stuff! Thanks for the api change 👍

raftmsohani reviewed (approved):
LGTM!

raftmsohani commented on tdrs-backend/tdpservice/stts/models.py:
```diff
@@ -39,6 +39,14 @@ class EntityType(models.TextChoices):
+    def num_sections(self):
+        """The number of sections this STT submits."""
+        if self.filenames is None:
+            return 4
```
Only nit: maybe add constant for returning value

elipe17 committed:
- use constant
- use double quotes

ADPennington committed:
Merge branch 'develop' into 2217-only-view-submission-needed

ADPennington commented:
lgtm! @victoriaatraft @reitermb -- do we need a release note associated with this change? 


_tribe (does not submit section 4)_
![STT_tribe_p3](https://github.com/user-attachments/assets/c00f8208-9334-4817-8587-746f33e6d672)
_state that submits SSP and TANF but doesnt submit section 4_
![STT_universe_p2](https://github.com/user-attachments/assets/f717b1df-8c79-44a0-8fb5-ea6bbc36a712)
_state that submits SSP and TANF and submits section 4_
![STT_sample_p1](https://github.com/user-attachments/assets/2b6e9891-48f4-428d-a118-0ab12a539f18)
_territory that submits section 4_
![STT_sample_p4](https://github.com/user-attachments/assets/af8b18e1-6df5-496e-b20c-a0581f565b78)

victoriaatraft commented:
@ADPennington Unless stt's have been submitting sections they didn't need to in the past and they'd be confused as to why it's not there anymore, I see this as I see this as "corrective" enhancement to the experience that helps people submit what they need to and doesn't need a release note. 
cc @reitermb

ADPennington reviewed (approved):
🚀

elipe17 committed:
Merge branch 'develop' into 2217-only-view-submission-needed
